### PR TITLE
Add forward AD support for clamp.Tensor

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74205
* __->__ #74042
* #74017

Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

Differential Revision: [D34851567](https://our.internmc.facebook.com/intern/diff/D34851567)

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -717,6 +717,57 @@ std::tuple<at::Tensor, at::Tensor> clamp_backward_min_max(
...
+      const auto min_lte_max = min_p <= max_p;
+      const auto pred_min = self_lte_min.logical_and(min_lte_max);
+
+      const auto max_lt_min = min_lte_max.logical_not();
+      const auto self_gte_max = self_lt_max.logical_not();
+      const auto& pred_max = areAnyTensorSubclassLike({min_p}) ?
+        self_gte_max.logical_or(max_lt_min) :
+        self_gte_max.logical_or_(max_lt_min);
+
+      return where(pred_self, self_t, scalar_zero) + where(pred_min, min_t, scalar_zero) + where(pred_max, max_t, scalar_zero);
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:745]: As you pointed out in the existing formulas, this is not needed.
You can do `where(self_p < min_p, min_t, where(self_p > max_p, max_t, self_t))` right?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -717,6 +717,57 @@ std::tuple<at::Tensor, at::Tensor> clamp_backward_min_max(
...
+    }
+  } else if (min_p.defined()) {
+    return where(self_p > min_p, self_t, min_t);
+  } else if (max_p.defined()) {
+    return where(self_p < max_p, self_t, max_t);
+  } else {
+    return self_t; // is this correct?
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:767]: If the function is the identity, yes

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp]: Lol big simplification! I had to nest another where on the outside though to handle the min > max case.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -717,6 +717,57 @@ std::tuple<at::Tensor, at::Tensor> clamp_backward_min_max(
   return ret;
 }
 
+at::Tensor clamp_jvp(
+  const Tensor& self_p, const Tensor& self_t,
+  const Tensor& min_p, const Tensor& min_t,
+  const Tensor& max_p, const Tensor& max_t
+) {
+  if (min_p.defined() && max_p.defined()) {
+    // We assume self_p is more likely to be tensor-subclass-like than max_p or min_p
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp]: No longer needed!

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851567).

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**